### ISSUE TITLE
Pyright VSCode extension eagerly turns itself off if Pyright is enabled, even when not in use

### ISSUE THREAD

**Issue by felixm-stripe** [microsoft/pyright#9325]
In the VSCode extension, Pyright tries to detect if Pylance is running:

https://github.com/microsoft/pyright/blob/5537ec9bcfeffdac543735fdf94e6d80318b5513/packages/vscode-pyright/src/extension.ts#L59-L60

However, this is done by just checking if the extension is enabled. This means that even though Pylance might not be in use, e.g. by setting:

```json
{
    "python.languageServer": "Jedi",
}
```

or:

```json
{
    "python.languageServer": "None",
}
```

Pyright will turn itself off. 

## Desired behavior
If Pylance is present and enabled but not in use, don't disable Pyright type-checking.

## Rejected solutions
* Uninstall Pylance -- a lot of the ecosystem depends on the Python extension which in turn depends on Pylance. This makes it hard to manage the developer setup without manual intervention whenever something gets updated and forces a re-install of Pylance
* Disable Pylance -- this is again a manual intervention that we don't want to force on our engineers.
* Configuration -- there is no support to selectively disable an extension via configuration.

**Comment by erictraut** [comment]
This behavior is currently by design. It's something we could consider changing if there's a good reason to do so — and we can identify a technical solution.

Can you explain why you want to use the pyright extension if you are a pylance user? Pylance provides a superset of features provided by pyright. The pyright extension is offered primarily for users who are using VSCodium or other variants of VS Code where Pylance is not licensed to run. It sounds like you have an atypical usage scenario, so I'd like to understand more.

I'm not sure if there's a way for pyright's client (the code code in extension.ts) to determine whether pylance is "in use" at activation time.

**Comment by felixm-stripe** [comment]
Hi Eric!

We don't depend directly on Pylance, rather we depend on the `ms-python.python` plugin, which in turn depends on Pylance. This is due to a lot of other tools in the ecosystem having dependencies on `ms-python.python`.

Can the Pyright client read config during activation to see if the language server is set to something else as per my OP?

**Comment by erictraut** [comment]
The Microsoft Python extension (`ms-python.python`) doesn't depend on Pylance. It will *load* Pylance if `python.languageServer` is set to "Default" (or "Pylance") *and* the Pylance extension is installed. If you are using the official version of VS Code, then I recommend you use Pylance as your language server rather than using the Pyright VS Code extension. If you're using VSCodium or some other fork of VS Code, then you won't be able to install Pylance, and there should be no conflict. 

Can you clarify whether you are using the official version of VS Code? If so, is there a reason why you want to use the Pyright extension rather than Pylance?

> Can the Pyright client read config during activation to see if the language server is set to something else as per my OP?

I'm not sure. That would require some investigation.

**Comment by felixm-stripe** [comment]
Absolutely!

We do use the official version of VS Code.

We can't use Pylance because it is materially different than Pyright, and we cannot control which version of Pyright that Pylance uses. As such, CI (Pyright) and VSCode (Pylance) typically don't agree and this is problematic at our scale.

The `ms-python.python` extension states:

> The Python extension will automatically install the following extensions by default to provide the best Python development experience in VS Code:
>
> - [Pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance) - to provide performant Python language support

https://github.com/microsoft/vscode-python/blob/40b29bf2beac97b77540ebeb96236aac652b78fb/README.md?plain=1#L12-L14

**Comment by erictraut** [comment]
Ah, got it. Thanks for explaining.

I'll investigate whether the extension can read the config during activation.

**Comment by erictraut** [comment]
I don't see a straightforward way do do this. The configs are not accessible at activation time.

I think this is pretty specific to your use case, so I'm going to close the enhancement request for now. I'm open to investing more time investigating solutions in the future if there's signal from other pyright users (e.g. thumbs-up reactions to this issue). 

If you'd like to explore potential solutions, you're welcome to do so. If you find a solution that meets your needs and is not too kludgey or difficult to maintain, I'd accept a PR.

**Comment by debonte** [comment]
> I don't see a straightforward way do do this. The configs are not accessible at activation time.

For what it's worth, Pylance obtains the values of a couple config settings in the extension when we're about to create the `LanguageClient`. We obtain the setting values via the `workspace.getConfiguration` vscode API and then pass the resulting values to the server via the `initializationOptions` property on `LanguageClientOptions`.

**Comment by felixm-stripe** [comment]
Thanks for the quick turnaround!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.387.
